pkg/credential/credential.go

```diff
@@ -39,53 +36,56 @@ func NewK8SCredentialProvider(k8sclient k8sclient.Interface, gsTenantID string,
-	if err != nil {
-		return auth.ClientCredentialsConfig{}, "", "", microerror.Mask(err)
+	if azureCluster.Spec.IdentityRef == nil {
+		return auth.ClientCredentialsConfig{}, "", "", microerror.Maskf(identityRefNotSetError, "AzureCluster %s/%s didn't have the IdentityRef field set", azureCluster.Namespace, azureCluster.Name)
 	}
 
-	subscriptionID, err := valueFromSecret(secret, subscriptionIDKey)
+	azureClusterIdentity := &v1alpha3.AzureClusterIdentity{}
+	err = k.k8sclient.CtrlClient().Get(ctx, client.ObjectKey{Namespace: azureCluster.Spec.IdentityRef.Namespace, Name: azureCluster.Spec.IdentityRef.Name}, azureClusterIdentity)
 	if err != nil {
 		return auth.ClientCredentialsConfig{}, "", "", microerror.Mask(err)
 	}
 
-	partnerID, err := valueFromSecret(secret, partnerIDKey)
+	secret := &v1.Secret{}
+	err = k.k8sclient.CtrlClient().Get(ctx, client.ObjectKey{Namespace: azureClusterIdentity.Spec.ClientSecret.Namespace, Name: azureClusterIdentity.Spec.ClientSecret.Name}, secret)
 	if err != nil {
-		// No having Partner ID in the secret means that customer has not
-		// upgraded yet to use the Azure Partner Program. In that case we set a
-		// constant random generated GUID that we haven't registered with Azure.
-		// When all customers have migrated, we should error out instead.
-		partnerID = defaultAzureGUID
+		return auth.ClientCredentialsConfig{}, "", "", microerror.Mask(err)
 	}
 
-	if _, exists := secret.GetLabels()[label.SingleTenantSP]; exists || tenantID == k.gsTenantID {
-		// The tenant cluster resources will belong to a subscription that belongs to the same Tenant ID used for authentication.
-		k.logger.Debugf(ctx, "Azure subscription %#q belongs to the same tenant ID %#q that owns the service principal. Using single tenant authentication", subscriptionID, tenantID)
-		credentials := auth.NewClientCredentialsConfig(clientID, clientSecret, tenantID)
-		return credentials, subscriptionID, partnerID, nil
+	subscriptionID := azureCluster.Spec.SubscriptionID
+	clientID := azureClusterIdentity.Spec.ClientID
+	tenantID := azureClusterIdentity.Spec.TenantID
+	// TODO find a way to store the partnerID
+	partnerID := defaultAzureGUID
+
+	clientSecret, err := valueFromSecret(secret, clientSecretKey)
+	if err != nil {
+		return auth.ClientCredentialsConfig{}, "", "", microerror.Mask(err)
 	}
 
 	// The tenant cluster resources will belong to a subscription that belongs to a different Tenant ID than the one used for authentication.
-	k.logger.Debugf(ctx, "Azure subscription %#q belongs to the tenant ID %#q which is different than the Tenant ID %#q that owns the Service Principal. Using multi tenant authentication", subscriptionID, tenantID, k.gsTenantID)
-	credentials := auth.NewClientCredentialsConfig(clientID, clientSecret, k.gsTenantID)
-	credentials.AuxTenants = append(credentials.AuxTenants, tenantID)
+	credentials := auth.NewClientCredentialsConfig(clientID, clientSecret, tenantID)
+	if tenantID != k.gsTenantID {
```

Comment:
This is the problematic bit. If the "SP tenant ID" (passed as a parameter to azure operator and defined in `installations` repo) and the tenant ID stored in the secret (let's call it BYOC tenant) are different, I always try to add one of the two as `AuxTenants`. I have zero idea what that means and if it can fail and why

Reply:
AFAICS this PR in current setting doesn't yet decouple Azure client factory (`client` package) from credentials factory (`credentials` package). If that was done and they were integrated with the specific credential struct I proposed earlier, could you add this lost field in that one?

Reply:
didn't do the decoupling yet. Plus I expanded my knowledge about the tenant issue and I'd like to brainstorm a solution for that.